File: 5/lab5.cpp
#include <iostream>
using namespace std;

/* ---------------------- БЫСТРАЯ СОРТИРОВКА ---------------------- */

// Меняет местами два числа
void swap(int& a, int& b) {
    int temp = a;
    a = b;
    b = temp;
}

// Разделяет массив относительно опорного элемента
int partition(int arr[], int low, int high) {
    int pivot = arr[high];  // Берем последний элемент как опорный
    int i = low - 1;        // Индекс меньших элементов

    for (int j = low; j < high; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(arr[i], arr[j]);
        }
    }

    swap(arr[i + 1], arr[high]);  // Ставим опорный элемент на место
    return i + 1;
}

// Рекурсивная функция быстрой сортировки
void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);

        quickSort(arr, low, pi - 1);   // Сортируем левую часть
        quickSort(arr, pi + 1, high);  // Сортируем правую
    }
}

/* ---------------------- СОРТИРОВКА СЛИЯНИЕМ ---------------------- */

// Объединяет две отсортированные части массива
void merge(int arr[], int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;

    int L[n1], R[n2];

    for (int i = 0; i < n1; i++)
        L[i] = arr[left + i];
    for (int j = 0; j < n2; j++)
        R[j] = arr[mid + 1 + j];

    int i = 0, j = 0, k = left;

    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) arr[k++] = L[i++];
        else arr[k++] = R[j++];
    }

    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];
}

// Рекурсивно делит массив и вызывает слияние
void mergeSort(int arr[], int left, int right) {
    if (left < right) {
        int mid = (left + right) / 2;

        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

/* ---------------------- ИНТЕРПОЛЯЦИОННЫЙ ПОИСК ---------------------- */

// Возвращает индекс найденного элемента или -1, если его нет
int interpolationSearch(int arr[], int n, int x) {
    int low = 0, high = n - 1;

    while (low <= high && x >= arr[low] && x <= arr[high]) {
        if (arr[low] == arr[high]) {
            if (arr[low] == x) return low;
            return -1;
        }

        // Формула вычисления предполагаемой позиции
        int pos = low + ((double)(high - low) / (arr[high] - arr[low])) * (x - arr[low]);

        if (arr[pos] == x) return pos;
        if (arr[pos] < x) low = pos + 1;
        else high = pos - 1;
    }

    return -1;
}

/* ---------------------- ОСНОВНАЯ ПРОГРАММА ---------------------- */

int main() {
    // Тест для быстрой сортировки
    int arr1[] = {10, 7, 8, 9, 1, 5};
    int n1 = sizeof(arr1) / sizeof(arr1[0]);

    cout << "Исходный массив (быстрая сортировка): ";
    for (int x : arr1) cout << x << " ";
    cout << endl;

    quickSort(arr1, 0, n1 - 1);

    cout << "После быстрой сортировки: ";
    for (int x : arr1) cout << x << " ";
    cout << "\n\n";


    // Тест для сортировки слиянием
    int arr2[] = {38, 27, 43, 3, 9, 82, 10};
    int n2 = sizeof(arr2) / sizeof(arr2[0]);

    cout << "Исходный массив (сортировка слиянием): ";
    for (int x : arr2) cout << x << " ";
    cout << endl;

    mergeSort(arr2, 0, n2 - 1);

    cout << "После сортировки слиянием: ";
    for (int x : arr2) cout << x << " ";
    cout << "\n\n";


    // Тест для интерполяционного поиска
    int sortedArr[] = {10, 12, 13, 16, 18, 19, 20, 21, 22, 23, 24, 33, 35, 42, 47};
    int n3 = sizeof(sortedArr) / sizeof(sortedArr[0]);
    int target = 18;

    int index = interpolationSearch(sortedArr, n3, target);

    cout << "Интерполяционный поиск числа " << target << ": ";
    if (index != -1) cout << "найдено на позиции " << index << endl;
    else cout << "не найдено" << endl;

    return 0;
}

File: 5/lab5.py
# ---------------------- СОРТИРОВКА ВСТАВКАМИ ----------------------

def insertion_sort(arr):
    """Сортировка вставками"""
    for i in range(1, len(arr)):
        key = arr[i]         # Текущий элемент
        j = i - 1            # Индекс предыдущего элемента

        # Сдвигаем элементы, пока не найдём место для key
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key


# Тест сортировки вставками
arr1 = [12, 11, 13, 5, 6]
print("Исходный массив (вставками):", arr1)
insertion_sort(arr1)
print("После сортировки:", arr1)
print()


# ---------------------- СОРТИРОВКА ШЕЛЛА ----------------------

def shell_sort(arr):
    """Сортировка Шелла"""
    gap = len(arr) // 2  # Начальный шаг

    while gap > 0:
        for i in range(gap, len(arr)):
            current_value = arr[i]
            position = i

            # Сдвигаем элементы, если нарушен порядок
            while position >= gap and arr[position - gap] > current_value:
                arr[position] = arr[position - gap]
                position -= gap
            arr[position] = current_value
        gap //= 2


# Тест сортировки Шелла
arr2 = [64, 34, 25, 12, 22, 11, 90]
print("Исходный массив (Шелла):", arr2)
shell_sort(arr2)
print("После сортировки:", arr2)
print()


# ---------------------- БИНАРНЫЙ ПОИСК ----------------------

def binary_search(arr, x):
    """Бинарный поиск элемента x в отсортированном массиве arr"""
    low = 0
    high = len(arr) - 1

    while low <= high:
        mid = (low + high) // 2

        if arr[mid] < x:
            low = mid + 1
        elif arr[mid] > x:
            high = mid - 1
        else:
            return mid  # Элемент найден

    return -1  # Если элемент не найден


# Тест бинарного поиска
sorted_array = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
search_element = 11

print("Массив для бинарного поиска:", sorted_array)
result = binary_search(sorted_array, search_element)
if result != -1:
    print(f"Элемент {search_element} найден на индексе {result}.")
else:
    print(f"Элемент {search_element} не найден.")
print()


# ---------------------- ПОИСК ФИБОНАЧЧИ ----------------------

def fibonacci_search(arr, x):
    """Поиск Фибоначчи"""
    fib_m_minus_2 = 0  # F(m-2)
    fib_m_minus_1 = 1  # F(m-1)
    fib_M = fib_m_minus_1 + fib_m_minus_2  # F(m)

    # Ищем минимальное число Фибоначчи >= длине массива
    while fib_M < len(arr):
        fib_m_minus_2 = fib_m_minus_1
        fib_m_minus_1 = fib_M
        fib_M = fib_m_minus_1 + fib_m_minus_2

    offset = -1  # Смещение начала отрезка

    while fib_M > 1:
        i = min(offset + fib_m_minus_2, len(arr) - 1)

        if arr[i] < x:
            fib_M = fib_m_minus_1
            fib_m_minus_1 = fib_m_minus_2
            fib_m_minus_2 = fib_M - fib_m_minus_1
            offset = i
        elif arr[i] > x:
            fib_M = fib_m_minus_2
            fib_m_minus_1 = fib_m_minus_1 - fib_m_minus_2
            fib_m_minus_2 = fib_M - fib_m_minus_1
        else:
            return i

    if fib_m_minus_1 and offset + 1 < len(arr) and arr[offset + 1] == x:
        return offset + 1

    return -1


# Тест поиска Фибоначчи
data = [10, 22, 35, 40, 45, 50, 80, 82, 85, 90, 100]
element_to_find = 85

print("Массив для поиска Фибоначчи:", data)
index = fibonacci_search(data, element_to_find)

if index != -1:
    print(f"Элемент {element_to_find} найден на индексе {index}.")
else:
    print(f"Элемент {element_to_find} не найден.")
